@xen-orchestra/backups-cli/index.js

```diff
@@ -0,0 +1,201 @@
+const { curryRight, flatten } = require('lodash')
+const { pipe, promisifyAll } = require('promise-toolbox')
+const { DISK_TYPE_DIFFERENCING } = require('vhd-lib/dist/_constants')
+const Vhd = require('vhd-lib').default
+
+const fs = promisifyAll(require('fs'))
+const handler = require('@xen-orchestra/fs').getHandler({ url: 'file://' })
+
+// -----------------------------------------------------------------------------
+
+const asyncMap = curryRight((iterable, fn) =>
+  Promise.all(
+    Array.isArray(iterable) ? iterable.map(fn) : Array.from(iterable, fn)
+  )
+)
+
+const fpify = method => (...args) =>
+  function(thisArg) {
+    return method.apply(thisArg, args)
+  }
+
+const filter = fpify([].filter)
+const endsWith = fpify(''.endsWith)
+
+const readDir = path =>
+  fs.readdir(path).then(
+    entries => {
+      entries.forEach((entry, i) => {
+        entries[i] = `${path}/${entry}`
+      })
+
+      return entries
+    },
+    () => []
+  )
+
+// -----------------------------------------------------------------------------
+
+const listVhds = pipe([
+  vmDir => vmDir + '/vdis',
+  readDir,
+  asyncMap(readDir),
+  flatten,
+  asyncMap(readDir),
+  flatten,
+  filter(endsWith('.vhd')),
+])
+
+async function handleVm(vmDir) {
+  const vhds = new Set()
+  const vhdParents = { __proto__: null }
+
+  // remove broken VHDs
+  await asyncMap(await listVhds(vmDir), async path => {
+    try {
+      const vhd = new Vhd(handler, path)
+      await vhd.readHeaderAndFooter()
+      vhds.add(path)
+      if (vhd.footer.diskType === DISK_TYPE_DIFFERENCING) {
+        vhdParents[path] = resolve(basename(path), vhd.header.parentUnicodeName)
+      }
+    } catch (error) {
+      console.warn('Error while checking VHD', path)
+      console.warn('  ', error)
+      if (error != null && error.code === 'ERR_ASSERTION') {
+        console.warn('  deleting…')
+        console.warn('')
+        force && (await fs.unlink(path))
+      }
+    }
+  })
+
+  // remove VHDs with missing ancestors
+  {
+    const deletions = []
+
+    const check = (parent, child) => {
+      // no longer needs to be checked
+      delete vhdParents[child]
+
+      // check the ancestors first
+      const grandparent = vhdParents[parent]
+      if (grandparent !== undefined) {
+        check(grandparent, parent)
+      }
+
+      if (!vhds.has(parent)) {
+        vhds.delete(child)
+
+        console.warn('Error while checking VHD', child)
+        console.warn('  missing parent', parent)
+        console.warn('  deleting…')
+        console.warn('')
+        force && deletions.push(fs.unlink(child))
+      }
+    }
+
+    // > A property that is deleted before it has been visited will not be
+    // > visited later.
+    // >
+    // > -- https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/for...in#Deleted_added_or_modified_properties
+    for (const child in vhdParents) {
+      check(vhdParents[child], child)
+    }
+
+    await Promise.all(deletions)
```

Comment:
Why do you await them here?

Reply:
It allows me to keep the check synchronous and decoupling the waiting for the actions.